consensus/src/creation/creator.rs

```diff
@@ -4,85 +4,109 @@ use crate::{
 };
 use log::trace;
 
+#[derive(Clone)]
+struct UnitsCollector<H: Hasher> {
+    n_candidates: NodeCount,
+    candidates: NodeMap<H::Hash>,
+}
+
+impl<H: Hasher> UnitsCollector<H> {
+    pub fn new(n_members: NodeCount) -> Self {
+        Self {
+            n_candidates: NodeCount(0),
+            candidates: NodeMap::with_size(n_members),
+        }
+    }
+
+    pub fn add_unit(&mut self, unit: &Unit<H>) {
+        let pid = unit.creator();
+        let hash = unit.hash();
+
+        if self.candidates.get(pid).is_none() {
+            self.candidates.insert(pid, hash);
+            self.n_candidates += NodeCount(1);
+            trace!(target: "AlephBFT-creator", "Added a new unit: {:#?}.", unit);
+        }
+    }
+
+    pub fn can_create(&self, node_id: NodeIndex) -> Option<&NodeMap<H::Hash>> {
```

Comment:
this could return an Err with variants like: `NotEnoughParents` & `MissingOwnUnit`. wdyt?

Reply:
I can do that, but it's not like we can do anything with that information except logging it, so why bother with a custom error. :)

Reply:
I feel like it makes it easier to log it at the callee side (like @timorl suggested) plus we never know if we won't use it later:
NotEnoughParents -> waitForParents
MissingOwnUnit -> waitForOwnUnits :P.
Also it looks better to return result here rather than option IMHO :P

Reply:
I also prefer it to return something more descriptive than `Option`. Just not a big fun of breaking KISS unnecessarily. My counterargument: errors should be used for recoverable situations, not just logging + those obvious values are going to change, but it's creating useless new dependency on them + it's not an error if we asked too early creator, so `Result` is not the correct one here. Any better suggestions?

Reply:
You can change the name of the function, if you want this to look more like an error (like `try_create`). 
Alternatively, you can return one custom enum which has 3 options.

I'm also OK with the current version, but I think the code is going to evolve soon and more precise return types might actually help in the rewrite.

Reply:
I feel like two of the arguments you present contradict each other. This is exactly a recoverable error, because creation doesn't have to succeed in this context, we recover from it by logging and ignoring the error.

If we use errors, we can have one fewer logline, we can avoid using logging in `creator.rs` completely (for the price of using `Display`), and we can have more precise tests. Plus, when we rewrite unit creation these two errors will likely remain and we will just add a couple more, if anything.

Reply:
I've decided to use `anyhow::Result` with a non public error (used only for logging, not leaking anything internal to creator).